### ISSUE TITLE
"Can't get D3D12SerializeVersionedRootSignature" assert

### ISSUE THREAD

**Issue by bkaradzic** [baldurk/renderdoc#436]
I just tried latest RenderDoc with bgfx, and I'm getting crash on init. Assert that's getting hit is in d3d12_debug.cpp lines 1541 and 396.

You can reproduce it in bgfx by enabling this line:
https://github.com/bkaradzic/bgfx/blob/master/src/renderer_d3d12.cpp#L501

**Comment by baldurk** [comment]
Hmm, that error would only happen if [the function](https://msdn.microsoft.com/en-us/library/windows/desktop/mt709110(v=vs.85).aspx) is missing from d3d12.dll and GetProcAddress returns NULL, and I'm not sure how that's possible.

```cpp
  PFN_D3D12_SERIALIZE_VERSIONED_ROOT_SIGNATURE serializeRootSig =
      (PFN_D3D12_SERIALIZE_VERSIONED_ROOT_SIGNATURE)GetProcAddress(
          GetModuleHandleA("d3d12.dll"), "D3D12SerializeVersionedRootSignature");

  if(serializeRootSig == NULL)
  {
    RDCERR("Can't get D3D12SerializeVersionedRootSignature");
    return NULL;
  }
```

Can you check your d3d12.dll, does it have that function export?

**Comment by bkaradzic** [comment]
`This function was released with the Windows 10 Anniversary Update (14393) and supersedes D3D12SerializeRootSignature.`

Trying to figure out what my version is (winver it says 10586.679, but not sure if it's referring to the same version as above).

**Comment by baldurk** [comment]
Oh for fuck's sake :angry:. D3D12 is such a mess of an API. Sounds like you should have the function in your version but who knows, I guess there could be a bunch of reasons for it to break :(.

If I can't count on that function being implemented I'll need to have a fallback. The problem is, I need that function for correct replay (since without it there can be root signatures I'm unable to re-create and modify when needed). So essentially you'll get captures that may or may not work depending on whether you have this function with no good way of signalling it.

**Comment by zao** [comment]
There has been two major end-user upgrades of Windows 10 thus far. They have a friendly string name for end users, a YYMM version, and an actual version. `winver.exe` tells you the YYMM code and the actual version.
There is a [version history](https://support.microsoft.com/en-us/help/12387/windows-10-update-history) available on the Microsoft website.

RTM was 10240.
The November Update (1511) is 15086.
The Anniversary Update (1607) is 15393.
The next update is some sort of "Creative Update" or something similarly vague, due for February 2017 or something along those lines.
Inbetween all these there are Insider builds every now and then.

When it comes to D3D12 development, you're likely going to see developers tracking the current Insider build, as they break and fix the library continuously judging by the discussions on [DirectX-Graphics-Samples](https://github.com/Microsoft/DirectX-Graphics-Samples/).

**Comment by baldurk** [comment]
Right, the bottom line being I can't count on this function being present ever, so I need to try it and then fall back to something else that will be present.

**Comment by bkaradzic** [comment]
This is enterprise Win10. Actually it's fine that RenderDoc fails to initialize in case this function is not present, I assume at some point my Win10 will updated with new D3D12.

**Comment by baldurk** [comment]
I added a fallback to the function from the original D3D12. I think it should work as long as you don't try and open a capture that uses root signature 1.1 features (although from my limited reading, 1.0 seems strictly more conservative than 1.1, so maybe it'll still work).

I can't test it properly since I didn't have an old win10 to run on, but I tried with deliberately NULL'ing the new function and it seemed to work OK.
